crates/data/src/bin/hypha-data.rs

```diff
@@ -0,0 +1,334 @@
+                                    }
+                                }
+                                Err(e) => {
+                                    tracing::warn!(error=%e, "Failed to construct relay listen address");
+                                }
+                            }
+
+                            Ok(peer_id)
+                        }
+                        Err(e) => Err(e),
+                    }
+                }
+            })
+            .collect::<Vec<_>>(),
+    )
+    .await;
+
+    let gateway_peer_ids: Vec<_> = gateway_results
+        .into_iter()
+        .filter_map(|result| result.ok())
+        .collect();
+
+    if gateway_peer_ids.is_empty() {
+        return Err(miette::miette!("Failed to connect to any gateway"));
+    }
+
+    tracing::info!(gateway_ids = ?gateway_peer_ids, "Connected to gateway(s)");
+
+    // NOTE: Wait until DHT bootstrapping is done.
+    network.wait_for_bootstrap().await.into_diagnostic()?;
+
+    // Treat each file in the directory as a subset of the dataset.
+    // TODO: determine subset ordering by, e.g., a naming convention of the files in the dataset folder.
+    let dataset_path = config.dataset_path();
+    if !dataset_path.is_dir() {
+        return Err(miette::miette!("Dataset path is not a directory"));
+    }
+
+    let dataset_name = match dataset_path.file_name().and_then(|name| name.to_str()) {
+        Some(name) => Ok(name),
+        None => Err(miette::miette!("Dataset path is not a directory")),
+    }?;
+    let dataset_files = match dataset_path.read_dir() {
+        Ok(dir) => Ok(dir
+            .filter_map(|entry| entry.ok().map(|entry| entry.path()))
+            .collect::<Vec<_>>()),
+        Err(err) => Err(miette::miette!("Failed to read dataset directory: {}", err)),
+    }?;
+
+    // Announce our dataset
+    // TODO: We need to announce the number of subsets of this dataset as well.
+    // This is needed for schedulers to figure out which subsets to assign to workers in data-parallel training.
+    tracing::info!(dataset_name, "Announcing");
+    let _ = network.provide(dataset_name).await;
+
+    let stream_pulls = network.streams_pull().expect("an unregistered pull stream").for_each_concurrent(None, {
+        |(peer_id, resource, mut stream)| {
+            let dataset_files = dataset_files.clone();
+            async move {
+                tracing::info!(peer_id = %peer_id, dataset = resource.dataset, "Sending tensor to peer");
+
+                if dataset_name == resource.dataset.as_str() {
+                    // Use the provided offsets to select a subset of the available dataset files.
+                    if resource.index as usize > dataset_files.len() {
+                        tracing::warn!(peer_id = %peer_id, dataset = resource.dataset, "Invalid index for dataset");
+                        return;
+                    }
+                    let dataset_slice = &dataset_files[resource.index as usize];
```

Comment:
We need guard against empty dirs in https://github.com/hypha-space/hypha/blob/fd7efe6b8ecab246feff7f03b67d6421c9294b5b/crates/data/src/bin/hypha-data.rs#L212-L217

as we'd otherwise run into a panic here for index `0`

Reply:
Added a check for empty directory.